File: flint/strategy/vwap_reversion.py
"""VWAP reversion strategy.

Buys when price drops X% below VWAP, sells when it returns to VWAP.
Works best in range-bound markets where price oscillates around fair value.
"""
from __future__ import annotations

from typing import TYPE_CHECKING, Dict, List, Optional

import numpy as np

from ..models import Candle, Signal
from .base import Strategy

if TYPE_CHECKING:
    from ..execution.context import ExecutionContext


class VWAPReversionStrategy(Strategy):
    """Buy below VWAP, sell when price reverts back toward VWAP."""

    def __init__(self, period: int = 20, entry_pct: float = 2.0,
                 exit_pct: float = 0.5) -> None:
        self.period = period
        self.entry_pct = entry_pct
        self.exit_pct = exit_pct

    @property
    def name(self) -> str:
        return f"VWAPReversion({self.period}, {self.entry_pct}%)"

    def reset(self) -> None:
        pass

    @classmethod
    def parameters(cls) -> Dict[str, dict]:
        return {
            "period": {"type": "int", "low": 10, "high": 50, "default": 20},
            "entry_pct": {"type": "float", "low": 0.5, "high": 5.0, "default": 2.0},
            "exit_pct": {"type": "float", "low": 0.1, "high": 2.0, "default": 0.5},
        }

    def on_candle(self, candle: Candle, history: List[Candle],
                  ctx: Optional["ExecutionContext"] = None) -> Signal:
        if len(history) < self.period:
            return Signal.HOLD

        window = history[-self.period:]
        typical_prices = np.array([(c.high + c.low + c.close) / 3 for c in window])
        volumes = np.array([c.volume for c in window])

        total_volume = float(np.sum(volumes))
        if total_volume == 0:
            return Signal.HOLD

        vwap = float(np.sum(typical_prices * volumes) / total_volume)
        price = candle.close
        deviation_pct = ((price - vwap) / vwap) * 100

        if ctx is not None:
            pos = ctx.position(candle.market)
            if pos is None:
                if deviation_pct <= -self.entry_pct:
                    size = (ctx.account.cash * 0.9) / price
                    if size > 0:
                        from ..models import Side
                        ctx.market_order(candle.market, Side.LONG, size)
                elif deviation_pct >= self.entry_pct:
                    size = (ctx.account.cash * 0.9) / price
                    if size > 0:
                        from ..models import Side
                        ctx.market_order(candle.market, Side.SHORT, size)
            else:
                if abs(deviation_pct) <= self.exit_pct:
                    ctx.close_position(candle.market)
            return Signal.HOLD

        # v1 fallback
        if deviation_pct <= -self.entry_pct:
            return Signal.BUY
        elif deviation_pct >= self.entry_pct:
            return Signal.SELL
        return Signal.HOLD

File: ui/src/templates/vwap-reversion.ts
export const VWAP_REVERSION_TEMPLATE = `from flint.strategy.base import Strategy
from flint.models import Candle, Signal, Side, TimeInForce
from typing import List
import numpy as np


class VWAPReversion(Strategy):
    """VWAP Reversion — buy below VWAP, sell on reversion.

    Uses v2 execution with stop-loss, impact checks, proper sizing.
    """
    def __init__(self, period=20, entry_pct=2.0, exit_pct=0.5, stop_pct=3.0):
        self.period = period
        self.entry_pct = entry_pct / 100
        self.exit_pct = exit_pct / 100
        self.stop_pct = stop_pct / 100

    @property
    def name(self): return f"VWAPReversion({self.period})"

    @classmethod
    def parameters(cls):
        return {
            "period": {"type": "int", "low": 10, "high": 50, "default": 20},
            "entry_pct": {"type": "float", "low": 0.5, "high": 5.0, "default": 2.0},
            "exit_pct": {"type": "float", "low": 0.1, "high": 2.0, "default": 0.5},
            "stop_pct": {"type": "float", "low": 1.0, "high": 6.0, "default": 3.0},
        }

    def on_candle(self, candle, history, ctx=None):
        if ctx is None or len(history) < self.period:
            return Signal.HOLD
        window = history[-self.period:]
        vol_sum = sum(c.volume for c in window)
        if vol_sum == 0:
            return Signal.HOLD
        vwap = sum(c.close * c.volume for c in window) / vol_sum
        dev = (candle.close - vwap) / vwap

        if not ctx.positions:
            side = None
            if dev < -self.entry_pct:
                side = Side.LONG
            elif dev > self.entry_pct:
                side = Side.SHORT
            if side:
                size = (ctx.account.cash * 0.9) / candle.close
                impact = ctx.get_impact_price(candle.market, side, size)
                if impact and abs(impact - candle.close) / candle.close > 0.002:
                    return Signal.HOLD
                if size > 0:
                    ctx.market_order(candle.market, side, size)
                    stop_side = Side.SHORT if side == Side.LONG else Side.LONG
                    stop_price = candle.close * ((1 - self.stop_pct) if side == Side.LONG else (1 + self.stop_pct))
                    ctx.stop_order(candle.market, stop_side, size, stop_price)
        else:
            if abs(dev) < self.exit_pct:
                ctx.close_position(candle.market)
                ctx.cancel_all(candle.market)

        return Signal.HOLD

    def reset(self): pass
`
